### PR TITLE
Support self-signed certificate on outputs

### PR DESCRIPTION
**Pull request by belimawr**
## What does this PR do?
Adds `ssl.ca_trusted_fingerprint` option, if set to a HEX fingerprint of a root CA certificate, this certificate is added to
the trusted CAs (as if it was defined on `ssl.certificate_authorities`), then the SSL validation continues as normal.

This happens during the SSL handshake.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
~~- [ ] I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [x] Make sure the UX is correct

## How to test this PR locally
1. Run an Elasticsearch 8.0 Instance with self signed certs
2. Edit the Elasticsearch output configuration from your beat:
```
output.elasticsearch:
  # Array of hosts to connect to.
  hosts: ["https://<your IP address>:9200"]

  username: "elastic"
  password: "<your elastic use password>"
  ssl.ca_trusted_fingerprint: "<ES CA certificate fingerprint>"
```
3. Run your Beat

### CODE REVIEW

File: libbeat/common/transport/tlscommon/tls_config.go [hunk 1]
```diff
@@ -248,6 +251,25 @@ func makeVerifyServerConnection(cfg *TLSConfig) func(tls.ConnectionState) error
...
+			// Provided fingerprint should match at least one certificate before we continue.
+			if bytes.Compare(digest[0:], fingerprint) == 0 {
+				return nil
+			}
+		}
+	}
+
+	// if CASha256 is not defined, or we could not validate the fingerprint,
+	// proceed with the normal certificate validation
+
```
Comment (belimawr) [inline libbeat/common/transport/tlscommon/tls_config.go:272]: That's just a PoC to ensure the fingerprint can be validated, this is far from the final implementation.

File: libbeat/common/transport/tlscommon/tls_config.go [hunk 2]
```diff
@@ -186,14 +230,18 @@ func makeVerifyConnection(cfg *TLSConfig) func(tls.ConnectionState) error {
 	case VerifyStrict:
 		if len(cfg.CASha256) > 0 {
 			return func(cs tls.ConnectionState) error {
+				if cfg.ESCAFingerprint != "" {
+					if err := trustESRootCA(cfg, cs.PeerCertificates); err != nil {
+						return err
+					}
+				}
```
Comment (belimawr) [inline libbeat/common/transport/tlscommon/tls_config.go:237]: We probably should not do this here. If the user configured `ssl.verification_mode: strict` this means they already have all necessary certificates in place and we should should not touch their configuration.

File: libbeat/common/transport/tlscommon/tls_config.go [hunk 3]
```diff
@@ -151,10 +159,41 @@ func (c *TLSConfig) BuildServerConfig(host string) *tls.Config {
...
+					cfg.RootCAs = x509.NewCertPool()
+				}
+				cfg.RootCAs.AddCert(cert)
+
+				cfg.CASha256 = append(cfg.CASha256, Fingerprint(cert))
+			}
+		}
+	}
+
+	return nil
```
Comment (belimawr) [inline libbeat/common/transport/tlscommon/tls_config.go:185]: Should we return an error here?

File: libbeat/common/transport/tlscommon/tls_config.go [hunk 4]
```diff
@@ -151,10 +159,41 @@ func (c *TLSConfig) BuildServerConfig(host string) *tls.Config {
...
+	fingerprint, err := hex.DecodeString(cfg.ESCAFingerprint)
+	if err != nil {
+		return fmt.Errorf("decode fingerprint: %w", err)
+	}
+
+	for _, cert := range peerCerts {
+		// Compute digest for each certificate.
+		digest := sha256.Sum256(cert.Raw)
+
+		if bytes.Compare(digest[0:], fingerprint) == 0 {
```
Comment (kvch) [inline libbeat/common/transport/tlscommon/tls_config.go:172]: This is a bit clearer:

```suggestion
		if bytes.Equal(digest[0:], fingerprint) {
```

File: libbeat/common/transport/tlscommon/tls_config.go (hunk 2 above)
Comment (kvch) [reply libbeat/common/transport/tlscommon/tls_config.go:237]: `VerifyFull` and `VerifyStrict` is essentially the same. The only difference is that `VerifyFull` accepts hostnames in the CommonName section of certs. I think you did the right thing in the first place, adding self-signed certificates is correct.

File: libbeat/common/transport/tlscommon/tls_config.go (hunk 3 above)
Comment (kvch) [reply libbeat/common/transport/tlscommon/tls_config.go:185]: What would be the error? No matching fingerprint?

File: libbeat/common/transport/tlscommon/tls_config.go (hunk 4 above)
Comment (belimawr) [reply libbeat/common/transport/tlscommon/tls_config.go:172]: Done

**Comment by kvch** [comment]
Please add the new option to the documentation, too: https://github.com/elastic/beats/tree/master/libbeat/docs/shared-ssl-config.asciidoc

File: libbeat/common/transport/tlscommon/tls_config.go (hunk 2 above)
Comment (belimawr) [reply libbeat/common/transport/tlscommon/tls_config.go:238]: Cool

File: libbeat/common/transport/tlscommon/tls_config.go (hunk 3 above)
Comment (belimawr) [reply libbeat/common/transport/tlscommon/tls_config.go]: Yes, because this fingerprint is also acting as a pin. I've updated the code to return an error.

**Comment by belimawr** [comment]
rebased onto master, force push.

**Review by kvch — APPROVED** [review]

**Pull request merged**